File: test/src/cpp/TestSpread.cc
#include "RPC.h"
#include <unistd.h>
#include <pthread.h>

using namespace moca::rpc;
int64_t authRequestId = 0;

void eventListener(const RPCClient *channel, int32_t eventType, RPCOpaqueData eventData, RPCOpaqueData userData)
{
  StringLite remoteAddress;
  uint16_t remotePort;
  channel->remoteAddress(&remoteAddress, &remotePort);
  StringLite localAddress;
  uint16_t localPort;
  channel->localAddress(&localAddress, &localPort);
  StringLite remoteId;
  channel->remoteId(&remoteId);
  StringLite localId;
  channel->localId(&localId);
  switch (eventType) {
    case EVENT_TYPE_CONNECTED:
      printf("Connected to server %s@%s:%u from %s@%s:%u\n", remoteId.str(), remoteAddress.str(), remotePort, localId.str(), localAddress.str(), localPort);
      break;
    case EVENT_TYPE_ESTABLISHED:
      {
        printf("Session to server %s@%s:%u from %s@%s:%u is established\n", remoteId.str(), remoteAddress.str(), remotePort, localId.str(), localAddress.str(), localPort);
        KeyValuePairs<StringLite, StringLite> headers;
        headers.append("u", "1");
        headers.append("s", "s");
        headers.append("k", "k");
        channel->request(&authRequestId, 1, &headers);
      }
      break;
    case EVENT_TYPE_DISCONNECTED:
      printf("Disconnected from server\n");
      break;
    case EVENT_TYPE_REQUEST:
      {
        RequestEventData *data = static_cast<RequestEventData *>(eventData);
        printf("Request %lld from server %s@%s:%u\n", data->id, remoteId.str(), remoteAddress.str(), remotePort);
        printf("Code: %d\n", data->code);
        for (size_t idx = 0, size = data->headers->size(); idx < size; ++idx) {
          const KeyValuePair<StringLite, StringLite> *pair = data->headers->get(idx);
          printf("Header %s => %s\n", pair->key.str(), pair->value.str());
        }
        printf("%d bytes payload\n", data->payloadSize);
        if (data->code == 4) {
          printf("Broadcast\n");
        } else {
          printf("Publish\n");
        }
      }
      break;
    case EVENT_TYPE_RESPONSE:
      {
        RequestEventData *data = static_cast<RequestEventData *>(eventData);
        printf("Response %lld from server %s@%s:%u\n", data->id, remoteId.str(), remoteAddress.str(), remotePort);
        printf("Code: %d\n", data->code);
        for (size_t idx = 0, size = data->headers->size(); idx < size; ++idx) {
          const KeyValuePair<StringLite, StringLite> *pair = data->headers->get(idx);
          printf("Header %s => %s\n", pair->key.str(), pair->value.str());
        }
        printf("%d bytes payload\n", data->payloadSize);
        if (authRequestId == data->id) {
          KeyValuePairs<StringLite, StringLite> headers;
          headers.append("s", "0:10:21");
          headers.append("f", "");
          int64_t tmp;
          channel->request(&tmp, 2, &headers);
        }
      }
      break;
    case EVENT_TYPE_PAYLOAD:
      {
        PayloadEventData *data = static_cast<PayloadEventData *>(eventData);
        printf("Payload of request %lld from server %s@%s:%u\n", data->id, remoteId.str(), remoteAddress.str(), remotePort);
        printf("Size : %d\n", data->size);
        printf("Payload : %lld\n", (*(int64_t *) data->payload));
        printf("Commit : %s\n", data->commit ? "true" : "false");
      }
      break;
    case EVENT_TYPE_ERROR:
      {
        ErrorEventData *data = static_cast<ErrorEventData *>(eventData);
        printf("Error %d:%s\n", data->code, data->message.str());
      }
      break;
  }
}

int main(int argc, char **argv)
{
  RPCClient *client = RPCClient::create(1000000000l);
  if (!client) {
    printf("Could not create\n");
    return 1;
  }
  client->addListener(eventListener);
  if (MOCA_RPC_FAILED(client->connect("127.0.0.1:1234"))) {
    printf("Could not connect\n");
    return 1;
  }
  while (true) {
    client->loop();
  }
  return 0;
}

File: test/src/cpp/TestSpread.m
#import "MCRPC.h"
#import <pthread.h>
#import <unistd.h>

@interface EventDelegate : NSObject<MCRPCEventDelegate>
@property (nonatomic, strong) NSString *remoteAddress;
@property (nonatomic, assign) uint16_t remotePort;
@property (nonatomic, strong) NSString *localAddress;
@property (nonatomic, assign) uint16_t localPort;
@property (nonatomic, strong) NSString *remoteId;
@property (nonatomic, strong) NSString *localId;
- (void) extract:(MCRPC *)channel;
- (void) onConnected:(MCRPC *)channel;
- (void) onEstablished:(MCRPC *)channel;
- (void) onDisconnected:(MCRPC *)channel;
- (void) onRequest:(MCRPC *)channel id:(int64_t)id code:(int32_t)code payloadSize:(int32_t)payloadSize headers:(NSArray *)headers;
- (void) onResponse:(MCRPC *)channel id:(int64_t)id code:(int32_t)code payloadSize:(int32_t)payloadSize headers:(NSArray *)headers;
- (void) onPayload:(MCRPC *)channel id:(int64_t)id commit:(bool)commit payload:(const void *)payload payloadSize:(int32_t)payloadSize;
- (void) onError:(MCRPC *)channel code:(int32_t)code message:(NSString *)message;
@end

@implementation EventDelegate
- (void) extract:(MCRPC *)channel
{
  NSString *address = NULL;
  uint16_t port;
  [channel remoteAddress:&address port:&port];
  self.remoteAddress = address;
  self.remotePort = port;
  [channel localAddress:&address port:&port];
  self.localAddress = address;
  self.localPort = port;
  NSString *id = NULL;
  [channel remoteId:&id];
  self.remoteId = id;
  [channel localId:&id];
  self.localId = id;
}

- (void) onConnected:(MCRPC *)channel
{
  [self extract:channel];
  NSLog(@"Connected to server %@@%@:%hu from %@@%@:%hu", self.remoteId, self.remoteAddress, self.remotePort, self.localId, self.localAddress, self.localPort);
}

- (void) onEstablished:(MCRPC *)channel
{
  [self extract:channel];
  NSLog(@"Session to server %@@%@:%hu from %@@%@:%hu is established", self.remoteId, self.remoteAddress, self.remotePort, self.localId, self.localAddress, self.localPort);
  [channel request:2 headers:@[[MCKeyValuePair alloc:@"s" value:@"t.0"], [MCKeyValuePair alloc:@"s" value:@"t."], [MCKeyValuePair alloc:@"s" value:@"t.010"],
         [MCKeyValuePair alloc:@"s" value:@"w.0"], [MCKeyValuePair alloc:@"s" value:@"w."], [MCKeyValuePair alloc:@"s" value:@"w.010"], [MCKeyValuePair alloc:@"s" value:@"cc"], 
         [MCKeyValuePair alloc:@"s" value:@"c.0"], [MCKeyValuePair alloc:@"s" value:@"c."], [MCKeyValuePair alloc:@"s" value:@"c.010"], [MCKeyValuePair alloc:@"s" value:@"cfg"], 
         [MCKeyValuePair alloc:@"s" value:@"l.0"], [MCKeyValuePair alloc:@"s" value:@"l.1"], [MCKeyValuePair alloc:@"s" value:@"l.010"]] payload:NULL payloadSize:0];
  [channel request:8 headers:@[[MCKeyValuePair alloc:@"s" value:@"ganging"]] payload:NULL payloadSize:0];
}

- (void) onDisconnected:(MCRPC *)channel
{
  [self extract:channel];
  NSLog(@"Disconnected from server");
}

- (void) onRequest:(MCRPC *)channel id:(int64_t)id code:(int32_t)code payloadSize:(int32_t)payloadSize headers:(NSArray *)headers
{
  [self extract:channel];
  NSLog(@"Request %lld from server %@@%@:%hu", id, self.remoteId, self.remoteAddress, self.remotePort);
  NSLog(@"Code: %d", code);
  NSLog(@"Headers: %@", headers);
  NSLog(@"%d bytes payload", payloadSize);
  //[channel response:id code:(code - 100) headers:headers payload:NULL payloadSize:0];
  if (code == 7) {
    NSLog(@"Session Push");
  } else {
    NSLog(@"Publish");
  }
}

- (void) onResponse:(MCRPC *)channel id:(int64_t)id code:(int32_t)code payloadSize:(int32_t)payloadSize headers:(NSArray *)headers
{
  [self extract:channel];
  NSLog(@"Response %lld from server %@@%@:%hu", id, self.remoteId, self.remoteAddress, self.remotePort);
  NSLog(@"Code: %d", code);
  NSLog(@"Headers: %@", headers);
  NSLog(@"%d bytes payload", payloadSize);
  //[channel request:8 headers:@[[MCKeyValuePair alloc:@"s" value:@"ganging"]] payload:NULL payloadSize:0];
}

- (void) onPayload:(MCRPC *)channel id:(int64_t)id commit:(bool)commit payload:(const void *)payload payloadSize:(int32_t)payloadSize
{
  [self extract:channel];
  NSLog(@"Payload of request %lld from server %@@%@:%hu", id, self.remoteId, self.remoteAddress, self.remotePort);
  NSLog(@"Size : %d", payloadSize);
  NSLog(@"Commit : %@", commit ? @"true" : @"false");
}

- (void) onError:(MCRPC *)channel code:(int32_t)code message:(NSString *)message;
{
  NSLog(@"Error %d:%@", code, message);
}

@end

void *thread(void *arg)
{
  EventDelegate *delegate = [[EventDelegate alloc] init];
  MCRPC *channel = [[MCRPC alloc] init:delegate];
  if (!channel) {
    NSLog(@"Could not create");
    return NULL;
  }
  if ([channel connect:[NSString stringWithUTF8String:(const char *) arg]] != 0) {
    NSLog(@"Could not connect");
    return NULL;
  }
  int32_t st;
  do {
    st = [channel loop];
  } while (st == MC_RPC_OK);
  return NULL;
}

int main(int argc, const char **argv)
{
  if (argc < 2) {
    NSLog(@"Missing required argument");
    return 1;
  }
  int numThreads = 1;
  pthread_t threads[numThreads]; 
  for (int idx = 0; idx < numThreads; ++idx) {
    pthread_create(threads + idx, NULL, thread, (void *) argv[1]);
  }

  for (int idx = 0; idx < numThreads; ++idx) {
    void *tmp;
    pthread_join(threads[idx], &tmp);
  }
  return 0;
}
